cmd/thanos/flags.go

```diff
@@ -36,6 +36,16 @@ func regCommonServerFlags(cmd *kingpin.CmdClause) (*string, *string, func(log.Lo
 
 	refreshInterval := cmd.Flag("cluster.refresh-interval", "Interval for membership to refresh cluster.peers state, 0 disables refresh.").Default(cluster.DefaultRefreshInterval.String()).Duration()
 
+	secretKey := cmd.Flag("cluster.secret-key", "Initial secret key to communicate cluster with encryption. Could be AES-128, AES-192, or AES-256 (hexadecimal data only)").HexBytes()
```

Comment:
How would the key be rotated in a running system?

Reply:
Probably, in no way, current implementation doesn't allow update primary key

>// After creating a Memberlist, the configuration given should not be
// modified by the user anymore.

but, I think, it could be possible call `peer.Join` next time with spare(rotated) key on the failure of message decoding with old key 🤔

Reply:
The 'serf' library (which wraps memberlist to provide more functionality) seems to provide a higher-level API for managing keys:
https://godoc.org/github.com/hashicorp/serf/serf#KeyManager

It might be helpful to see how Serf (or Consul, which uses Serf) is doing key rotation.

Reply:
In the memberlist library, there's:
https://godoc.org/github.com/hashicorp/memberlist#Keyring

But it lacks any orchestratration, e.g. how to determine when a new key has propagated to all peers (so you can stop encrypting messages with the old key). The Serf library seems to provide this using the key manager.

We also need to figure out how we pass the new key (maybe we pass the old and new keys as commandline-flags). Consul does it using a CLI command:
https://www.consul.io/docs/commands/keyring.html

Reply:
well, I would suggest to move research of using the Serf library in another PR, as it requires a bit more changes, than I thought.

As it turned out, for the key rotation, it uses a separate file https://github.com/hashicorp/serf/blob/master/serf/serf.go#L1706, which means all components have to be a stateful service